core/base/mergeTreePrincipalGeodesics/MergeTreePrincipalGeodesics.h

```diff
@@ -0,0 +1,1652 @@
+      double v2_norm = ttk::Geometry::magnitude(v2_flatten);
+      double beta = v2_norm / (v1_norm + v2_norm);
+      std::vector<double> v;
+      ttk::Geometry::addVectors(v1_flatten, v2_flatten, v);
+      ttk::Geometry::scaleVector(v, (1 - beta), v1_flatten);
+      ttk::Geometry::scaleVector(v, beta, v2_flatten);
+      ttk::Geometry::unflattenMultiDimensionalVector(v1_flatten, v1, 2);
+      ttk::Geometry::unflattenMultiDimensionalVector(v2_flatten, v2, 2);
+    }
+
+    void
+      orthogonalProjection(std::vector<std::vector<double>> &v1,
+                           std::vector<std::vector<double>> &v2,
+                           std::vector<std::vector<std::vector<double>>> &vS,
+                           std::vector<std::vector<std::vector<double>>> &v2s) {
+      // Multi flatten and sum vS and v2s
+      std::vector<std::vector<double>> sumVs;
+      ttk::Geometry::multiAddVectorsFlatten(vS, v2s, sumVs);
+
+      // Flatten v1 and v2
+      std::vector<double> v1_flatten, v2_flatten, v1_proj, v2_proj;
+      ttk::Geometry::flattenMultiDimensionalVector(v1, v1_flatten);
+      ttk::Geometry::flattenMultiDimensionalVector(v2, v2_flatten);
+
+      // Call Gram Schmidt
+      callGramSchmidt(sumVs, v1_flatten, v1_proj);
+      callGramSchmidt(sumVs, v2_flatten, v2_proj);
+
+      // Unflatten the resulting vectors
+      ttk::Geometry::unflattenMultiDimensionalVector(v1_proj, v1, 2);
+      ttk::Geometry::unflattenMultiDimensionalVector(v2_proj, v2, 2);
+    }
+
+    // TODO avoid copying vectors
+    template <class dataType>
+    double
+      projectionStep(int geodesicNumber,
+                     ftm::MergeTree<dataType> &barycenter,
+                     std::vector<std::vector<double>> &v,
+                     std::vector<std::vector<double>> &v2,
+                     std::vector<std::vector<std::vector<double>>> &vS,
+                     std::vector<std::vector<std::vector<double>>> &v2s,
+                     ftm::MergeTree<dataType> &barycenter2,
+                     std::vector<std::vector<double>> &trees2V,
+                     std::vector<std::vector<double>> &trees2V2,
+                     std::vector<std::vector<std::vector<double>>> &trees2Vs,
+                     std::vector<std::vector<std::vector<double>>> &trees2V2s,
+                     bool useSecondInput,
+                     unsigned int noProjectionStep) {
+      std::vector<std::vector<std::vector<double>>> vSConcat, v2sConcat;
+      if(useSecondInput) {
+        Timer t_vectorCopy;
+        vSConcat = vS;
+        v2sConcat = v2s;
+        for(unsigned int j = 0; j < vS.size(); ++j) {
+          vSConcat[j].insert(
+            vSConcat[j].end(), trees2Vs[j].begin(), trees2Vs[j].end());
+          v2sConcat[j].insert(
+            v2sConcat[j].end(), trees2V2s[j].begin(), trees2V2s[j].end());
+        }
+        t_vectorCopy_time_ += t_vectorCopy.getElapsedTime();
+      }
+
+      // double oldCost = -1, oldCost2 = -1;
```

Comment:
thanks for removing commented code.
thanks for double-checking that in other places of your code.

Reply:
I did another check for removing old code in the next commit, thanks!